src/internal/analytics/__tests__/components/analytics-funnel.test.tsx

```diff
@@ -0,0 +1,346 @@
+
+jest.useFakeTimers(); // Mock timers
+jest.mock('../../../analytics', () => ({
+  __esModule: true,
+  default: {
+    funnelStart: jest.fn(() => mockedFunnelInteractionId),
+    funnelComplete: jest.fn(),
+    funnelCancelled: jest.fn(),
+    funnelStepStart: jest.fn(),
+    funnelStepComplete: jest.fn(),
+    funnelSubStepStart: jest.fn(),
+    funnelSubStepComplete: jest.fn(),
+  },
+}));
+
+describe('AnalyticsFunnel', () => {
+  beforeEach(() => {
+    jest.runAllTimers(); // Advance the timers to execute the setTimeout callback
+    jest.clearAllMocks(); // Reset all mock function call counters
+  });
+
+  test('renders children', () => {
+    const { getByText } = render(
+      <AnalyticsFunnel funnelType={FunnelType.SINGLE_PAGE} optionalStepNumbers={[]} totalFunnelSteps={3}>
+        <div>Child Content</div>
+      </AnalyticsFunnel>
+    );
+
+    const childElement = getByText('Child Content');
+    expect(childElement).toBeInTheDocument();
+  });
+
+  test('adds analytics data attributes to root node', () => {
+    const ChildComponent = () => {
+      const { funnelProps } = useFunnel();
+
+      return <div data-testid="root" {...funnelProps} />;
+    };
+
+    const funnelType = FunnelType.SINGLE_PAGE;
+    const optionalStepNumbers = [1, 3];
+    const totalFunnelSteps = 5;
+
+    const { getByTestId } = render(
+      <AnalyticsFunnel
+        funnelType={funnelType}
+        optionalStepNumbers={optionalStepNumbers}
+        totalFunnelSteps={totalFunnelSteps}
+      >
+        <ChildComponent />
+      </AnalyticsFunnel>
+    );
+
+    // Check if the root element has a specific data attribute
+    expect(getByTestId('root')).toHaveAttribute('data-analytics-funnel-interaction-id', expect.any(String));
+  });
+
+  test('calls funnelStart when the component renders', () => {
+    render(<AnalyticsFunnel funnelType={FunnelType.SINGLE_PAGE} optionalStepNumbers={[]} totalFunnelSteps={1} />);
+
+    expect(FunnelMetrics.funnelStart).toHaveBeenCalledTimes(1);
+    expect(FunnelMetrics.funnelStart).toHaveBeenCalledWith(
+      expect.objectContaining({
+        funnelNameSelector: expect.any(String),
+      })
+    );
+  });
+
+  test('calls funnelComplete when the component unmounts after submitting', () => {
+    // ChildComponent is a sample component that renders a button to call funnelSubmit
+    const ChildComponent = () => {
+      const { funnelSubmit } = useFunnel();
+
+      return <button onClick={funnelSubmit}>Submit</button>;
+    };
+
+    const { unmount, getByText } = render(
+      <AnalyticsFunnel funnelType={FunnelType.SINGLE_PAGE} optionalStepNumbers={[]} totalFunnelSteps={3}>
+        <ChildComponent />
+      </AnalyticsFunnel>
+    );
+    expect(FunnelMetrics.funnelComplete).not.toHaveBeenCalled();
+
+    fireEvent.click(getByText('Submit')); // Trigger the button click event
+    unmount();
+    jest.runOnlyPendingTimers(); // We use setTimeout on unmount to fire funnelComplete after the current event loop
```

Comment:
I have got the point of replacing 7000 timeout in #1104, but using fake timers for shorter timeouts is not very useful. 

[waitFor](https://testing-library.com/docs/dom-testing-library/api-async/#waitfor) would be more appropriate here

Reply:
Why is waitFor more appropriate here? The way I understand them is that waitFor is intended for waiting for a particular state with promises, usually async fetching of data, and jest fake timers are intended for when you have setTimeout, setInterval etc and you want to control the behavior.

Reply:
Because we try to write tests not too much tied to implementation details (react-testing-library has been promoting this approach for years).

Reply:
Changed to waitFor in next revision